Refactor documentation and update reference manual

jbcrail opened the pull request:
This has several updates, including:

- refreshed the Sphinx build and configuration
- updated the directory structure to adhere to conventions used by Dask
- fixed several Sphinx warnings, spelling errors, and deprecated examples
- simplified URL references
- added instructions/requirements needed to build the docs locally
- added new methods to the API documentation

Related to #237

jbednar commented:
Note that a lot of those links are from notebooks not on github, e.g. Anaconda.org

jbcrail commented:
It wasn't necessary to move the content. When I re-ran `sphinx-quickstart`, it used `source` by default. I didn't realize the notebook examples used the location. I reverted back to the original location.

I also updated the notebook links in the index to refer to GitHub instead of Anaconda.org.
